nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java

```diff
@@ -0,0 +1,585 @@
+
+                        logger.error(message);
+                        if (logger.isDebugEnabled()) {
+                            logger.error(message, t);
+                        }
+                    }
+                }
+            }, 30, 10, TimeUnit.SECONDS);
+        }
+
+        descriptors = new ArrayList<>();
+
+        // The following is required for JRuby, should be transparent to everything else.
+        // Note this is not done in a ScriptEngineConfigurator, as it is too early in the lifecycle. The
+        // setting must be there before the factories/engines are loaded.
+        System.setProperty("org.jruby.embed.localvariable.behavior", "persistent");
+
+        // Create list of available engines
+        ScriptEngineManager scriptEngineManager = new ScriptEngineManager();
+        List<ScriptEngineFactory> scriptEngineFactories = scriptEngineManager.getEngineFactories();
+        if (scriptEngineFactories != null) {
+            scriptEngineFactoryMap = new HashMap<>(scriptEngineFactories.size());
+            List<AllowableValue> engineList = new LinkedList<>();
+            for (ScriptEngineFactory factory : scriptEngineFactories) {
+                engineList.add(new AllowableValue(factory.getLanguageName()));
+                scriptEngineFactoryMap.put(factory.getLanguageName(), factory);
+            }
+
+            // Sort the list by name so the list always looks the same.
+            Collections.sort(engineList, new Comparator<AllowableValue>() {
+                @Override
+                public int compare(AllowableValue o1, AllowableValue o2) {
+                    if (o1 == null) {
+                        return o2 == null ? 0 : 1;
+                    }
+                    if (o2 == null) {
+                        return -1;
+                    }
+                    return o1.getValue().compareTo(o2.getValue());
+                }
+            });
+
+            AllowableValue[] engines = engineList.toArray(new AllowableValue[engineList.size()]);
+
+            SCRIPT_ENGINE = new PropertyDescriptor.Builder()
+                    .name("Script Engine")
+                    .required(true)
+                    .description("The engine to execute scripts")
+                    .allowableValues(engines)
+                    .defaultValue(engines[0].getValue())
+                    .required(true)
+                    .expressionLanguageSupported(false)
+                    .build();
+            descriptors.add(SCRIPT_ENGINE);
+        }
+
+        descriptors.add(SCRIPT_FILE);
+        descriptors.add(SCRIPT_BODY);
+        descriptors.add(SCRIPT_ARGS);
+        descriptors.add(MODULES);
+
+        isInitialized.set(true);
+    }
+
+    /**
+     * Returns the valid relationships for this processor.
+     *
+     * @return a Set of Relationships supported by this processor
+     */
+    @Override
+    public Set<Relationship> getRelationships() {
```

Comment:
This is related to my previous comment.  We cannot really know whether they'll operate on content or not and how many flow files and so on.  I believe we should simply give them the session and let them have at it.  Tell them you have two options "the flow files you transfer are going to success" or 'failure' as those are the only two options.  Orig makes sense if we enforce they will only have access to a single source flow file but since we're giving them the session it is a bit different.